File: web/src/lib/passwordPolicy.js
// Refleja PasswordPolicy del backend (que es la AUTORIDAD; esto es solo la ayuda visual y la
// validación previa al envío). Las reglas usan las MISMAS clases ASCII que el backend
// ([A-Z], [a-z], [0-9], [^A-Za-z0-9]) para que back y front nunca discrepen.
export const PASSWORD_RULES = [
  { key: 'length', label: 'Al menos 8 caracteres', test: (p) => p.length >= 8 },
  { key: 'upper', label: 'Una letra mayúscula (A-Z)', test: (p) => /[A-Z]/.test(p) },
  { key: 'lower', label: 'Una letra minúscula (a-z)', test: (p) => /[a-z]/.test(p) },
  { key: 'digit', label: 'Un número (0-9)', test: (p) => /[0-9]/.test(p) },
  { key: 'special', label: 'Un carácter especial (!@#$…)', test: (p) => /[^A-Za-z0-9]/.test(p) },
]

// true si la contraseña cumple TODAS las reglas.
export function isPasswordStrong(password) {
  const p = password || ''
  return PASSWORD_RULES.every((r) => r.test(p))
}

File: backend/ArquaBilling.Api/Common/PasswordPolicy.cs
namespace ArquaBilling.Api.Common;

// FUENTE ÚNICA de las reglas de fortaleza de contraseña. Se usa al ESTABLECER una
// contraseña: crear usuario, restablecer contraseña y sembrar el admin de producción.
//
// NO aplica al LOGIN: un usuario con una contraseña vieja "débil" debe poder seguir
// entrando; la política solo filtra contraseñas NUEVAS.
//
// Los checks son ASCII a propósito, para que coincidan EXACTO con los del frontend
// (regex [A-Z], [a-z], [0-9], [^A-Za-z0-9]) — la doble red no debe discrepar.
public static class PasswordPolicy
{
    public const int MinLength = 8;

    // Devuelve la lista de reglas NO cumplidas (vacía = contraseña válida). Cada mensaje es
    // claro y en español, listo para mapear por campo en el envelope { message, details }.
    public static IReadOnlyList<string> Validate(string? password)
    {
        var p = password ?? string.Empty;
        var errors = new List<string>();

        if (p.Length < MinLength)
        {
            errors.Add($"Debe tener al menos {MinLength} caracteres.");
        }
        if (!p.Any(char.IsAsciiLetterUpper))
        {
            errors.Add("Debe incluir al menos una letra mayúscula.");
        }
        if (!p.Any(char.IsAsciiLetterLower))
        {
            errors.Add("Debe incluir al menos una letra minúscula.");
        }
        if (!p.Any(char.IsAsciiDigit))
        {
            errors.Add("Debe incluir al menos un número.");
        }
        if (!p.Any(c => !char.IsAsciiLetterOrDigit(c)))
        {
            errors.Add("Debe incluir al menos un carácter especial.");
        }

        return errors;
    }

    public static bool IsValid(string? password) => Validate(password).Count == 0;
}
